Could not import faiss-node (NextJS) when loading from python

Hi, I know this is a duplicate of:

https://github.com/langchain-ai/langchainjs/issues/2107

However, in that issue it says that this has been resolved. While I'm on the latest release:
0.1.2

Which should have the fix for this available here:
https://github.com/langchain-ai/langchainjs/pull/2178

My issue is the following:
https://github.com/langchain-ai/langchainjs/issues/2107#issuecomment-1891106038

For reference, I'll post it below:

Hi there,

I'm getting the same type of error message in my console when I try to:
```javascript
const vectorstore = await FaissStore.loadFromPython("./knowledge", new OpenAIEmbeddings({
            modelName: "text-embedding-ada-002"
}));
```

knowledge is a folder in the root folder that includes 2 files:
```
index.pkl
index.faiss
```

I am using the following to see whether my API route in nextjs can reach the files:
```javascript
const knowledge = path.resolve('./knowledge');
const filenames = fs.readdirSync(knowledge);
console.log(filenames);
```
this logs the actual filenames, so I know they're there...

I'm getting the following error message in my console:
```
 Error: Could not import faiss-node. Please install faiss-node as a dependency with, e.g. `npm install -S faiss-node`.

Error: Cannot read properties of undefined (reading 'indexOf')
    at FaissStore.importFaiss (webpack-internal:///(rsc)/./node_modules/@langchain/community/dist/vectorstores/faiss.js:363:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async readIndex (webpack-internal:///(rsc)/./node_modules/@langchain/community/dist/vectorstores/faiss.js:292:37)   
    at async Promise.all (index 1)
    at async FaissStore.loadFromPython (webpack-internal:///(rsc)/./node_modules/@langchain/community/dist/vectorstores/faiss.js:295:32)
    at async POST (webpack-internal:///(rsc)/./app/api/chat/route.ts:56:29)
    at async D:\Development\Persoonlijk\gpt-creator\node_modules\next\dist\compiled\next-server\app-route.runtime.dev.js:6:63251

```

I'm running this on windows with ```npm run dev```
I've already tried installing the package, but I have no clue why my nextjs isn't importing the files...

My node version is 20.10.0


Is there anything I can do to get this issue sorted?

It says read properties of undefined (reading 'indexOf')
This to me indicates it can not read the ./knowledge folder with files for some reason. But I haven't been able to figure out why yet. I could be totally wrong here though.

@ewfian 

Where do I import FakeEmbeddings from?

The suggested packages by my IDE seem to all fail:
```
Module not found: Package path ./dist/utils/testing is not exported from package /home/simbaclaws/Projects/Personal/my-gpt/node_modules/@langchain/community (see exports field in /home/simbaclaws/Projects/Personal/my-gpt/node_modules/@langchain/community/package.json)
```

```
Module not found: Package path ./dist/embeddings/fake is not exported from package /home/simbaclaws/Projects/Personal/my-gpt/node_modules/langchain (see exports field in /home/simbaclaws/Projects/Personal/my-gpt/node_modules/langchain/package.json
```

> @ewfian
> 
> Where do I import FakeEmbeddings from?
> 
> The suggested packages by my IDE seem to all fail:
> 
> ```
> Module not found: Package path ./dist/utils/testing is not exported from package /home/simbaclaws/Projects/Personal/my-gpt/node_modules/@langchain/community (see exports field in /home/simbaclaws/Projects/Personal/my-gpt/node_modules/@langchain/community/package.json)
> ```
> 
> ```
> Module not found: Package path ./dist/embeddings/fake is not exported from package /home/simbaclaws/Projects/Personal/my-gpt/node_modules/langchain (see exports field in /home/simbaclaws/Projects/Personal/my-gpt/node_modules/langchain/package.json
> ```

"langchain/embeddings/fake"

--- a/libs/langchain-community/src/vectorstores/faiss.ts
+++ b/libs/langchain-community/src/vectorstores/faiss.ts
@@ -323,9 +323,15 @@ export class FaissStore extends SaveableVectorStore {
           "InMemoryDocstore",
           PyInMemoryDocstore
         )
+        .register(
+          "langchain_community.docstore.in_memory",
+          "InMemoryDocstore",
+          PyInMemoryDocstore
+        )
         .register("langchain.schema", "Document", PyDocument)
         .register("langchain.docstore.document", "Document", PyDocument)
         .register("langchain.schema.document", "Document", PyDocument)
+        .register("langchain_core.documents.base", "Document", PyDocument)
         .register("pathlib", "WindowsPath", (...args) => args.join("\\"))
         .register("pathlib", "PosixPath", (...args) => args.join("/"));
 

--- a/libs/langchain-community/src/vectorstores/tests/faiss.int.test.data/faiss.int.test.py
+++ b/libs/langchain-community/src/vectorstores/tests/faiss.int.test.data/faiss.int.test.py
@@ -1,10 +1,9 @@
 from langchain.embeddings.openai import OpenAIEmbeddings
 from langchain.text_splitter import CharacterTextSplitter
-from langchain.vectorstores import FAISS
-from langchain.document_loaders import TextLoader
+from langchain_community.vectorstores import FAISS
+from langchain_community.document_loaders import TextLoader
 
-from langchain.document_loaders import TextLoader
-loader = TextLoader('../../../../../examples/state_of_the_union.txt')
+loader = TextLoader('../../../../../../examples/state_of_the_union.txt')
 documents = loader.load()
 text_splitter = CharacterTextSplitter(chunk_size=1000, chunk_overlap=0)
 docs = text_splitter.split_documents(documents)

--- a/libs/langchain-community/src/vectorstores/tests/faiss.int.test.data/requirements.txt
+++ b/libs/langchain-community/src/vectorstores/tests/faiss.int.test.data/requirements.txt
@@ -1,1 +1,2 @@
-langchain==0.0.253
+langchain==0.1.0
+langchain-community==0.0.12

--- a/libs/langchain-community/src/vectorstores/tests/faiss.int.test.ts
+++ b/libs/langchain-community/src/vectorstores/tests/faiss.int.test.ts
@@ -101,7 +101,7 @@ test("Test FaissStore.loadFromPython", async () => {
   );
 
   const expectedResultofPythonSaved = new Document({
-    metadata: { source: "../../../../../examples/state_of_the_union.txt" },
+    metadata: { source: "../../../../../../examples/state_of_the_union.txt" },
     pageContent: `Tonight. I call on the Senate to: Pass the Freedom to Vote Act. Pass the John Lewis Voting Rights Act. And while you’re at it, pass the Disclose Act so Americans can know who is funding our elections. 
 
 Tonight, I’d like to honor someone who has dedicated his life to serve this country: Justice Stephen Breyer—an Army veteran, Constitutional scholar, and retiring Justice of the United States Supreme Court. Justice Breyer, thank you for your service. 

